### PR TITLE
Test datetime type in MariaDB connector

### LINKED ISSUE #17242: Add test for datetime type in MariaDB connector
Got new requirement for datetime support in MariaDB connector

### PR DESCRIPTION
**Pull request by abusk**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?
improvement
Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
MariaDB connector
How would you describe this change to a non-technical end user or system administrator?
Support datetime type in MariaDB connector

## Additional context and related issues
Fixes #17242 

## Release notes
(x) This is not user-visible or is docs only, and no release notes are required.

### CODE REVIEW

File: plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java [hunk 1]
```diff
@@ -329,6 +329,13 @@ public Optional<ColumnMapping> toColumnMapping(ConnectorSession session, Connect
             return unsignedMapping;
         }
 
+        String jdbcTypeName = typeHandle.getJdbcTypeName()
+                .orElseThrow(() -> new TrinoException(JDBC_ERROR, "Type name is missing: " + typeHandle));
+        if ("datetime".equalsIgnoreCase(jdbcTypeName)) {
+            TimestampType timestampType = createTimestampType(getTimestampPrecision(typeHandle.getRequiredColumnSize()));
+            return Optional.of(timestampColumnMapping(timestampType));
+        }
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java:337]: I don't understand why logic is requried. The new test (TestMariaDbTypeMapping#testTimestamp) passes without this change.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java [hunk 2]
```diff
@@ -812,6 +813,76 @@ private void testTimestamp(ZoneId sessionZone)
...
+    }
+
+    private void testDatetime(ZoneId sessionZone)
+    {
+        Session session = Session.builder(getSession())
+                .setTimeZoneKey(getTimeZoneKey(sessionZone.getId()))
+                .build();
+
+        SqlDataTypeTest.create()
+                .addRoundTrip("datetime(3)", "'1958-01-01 13:18:03.123'", createTimestampType(3), "TIMESTAMP '1958-01-01 13:18:03.123'")
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:833]: Please update `testTimestamp(ZoneId sessionZone)` method to accept both timestamp and datetime.

File: docs/src/main/sphinx/connector/mariadb.md [hunk 3]
```diff
@@ -261,6 +261,9 @@ to the following table:
       `explicit_defaults_for_timestamp
       <https://mariadb.com/docs/reference/mdb/system-variables/explicit_defaults_for_timestamp/>`_
       to avoid implicit default values and use `NULL` as the default value.
+* - `DATETIME(n)`
+  - `TIMESTAMP(n)`
+  -
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/mariadb.md:266]: This section is Trino → MariaDB type mapping. Please move to "MariaDB type to Trino type mapping" section.

**Comment by ebyhr** [comment]
> Added Support for datetime type in MariaDB connector

`datetime` is already supported as far as I know. Also, please avoid past tense in the commit title.
https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java [hunk 4]
```diff
@@ -112,7 +112,6 @@ protected Optional<DataMappingTestSetup> filterDataMappingSmokeTestData(DataMapp
         if (typeName.equals("timestamp(6)")) {
             return Optional.of(new DataMappingTestSetup("timestamp(6)", "TIMESTAMP '2020-02-12 15:03:00'", "TIMESTAMP '2038-01-19 03:14:07.000000'"));
         }
-
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java:115]: Revert unrelated change.

File: plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java (hunk 1 above)
Comment (abusk) [reply plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java:337]: `java.sql.Types` does not have `datatime` type, due to that can't map the `datetime`  based on jdbcType like it was done for `timestamp` inside the switch case condition.  For `datatime` we have to check based on the `jdbcTypeName` from the `typeHandle`.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java (hunk 2 above)
Comment (abusk) [reply plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:833]: `MariaDB stores values that use the DATETIME data type in a format that supports values between 1000-01-01 00:00:00.000000 and 9999-12-31 23:59:59.999999` which is different from TIMESTAMP (https://mariadb.com/kb/en/timestamp/#supported-values).  As the test cases has different ranges of values, so it is better to keep in separate.

File: docs/src/main/sphinx/connector/mariadb.md (hunk 3 above)
Comment (abusk) [reply docs/src/main/sphinx/connector/mariadb.md:266]: ah yea. Fixed.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java (hunk 4 above)
Comment (abusk) [reply plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java:115]: True. Reverted.

File: plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java (hunk 1 above)
Comment (ebyhr) [reply plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java:337]: Could you explain why the new test passes without this change?

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java (hunk 2 above)
Comment (ebyhr) [reply plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java]: Please leave the code comment.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java [hunk 5]
```diff
@@ -812,6 +813,76 @@ private void testTimestamp(ZoneId sessionZone)
                 .execute(getQueryRunner(), session, trinoCreateAndInsert("test_timestamp"));
     }
 
+    @Test
+    public void testDatetime()
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:817]: The current place is between `testTimestamp` and `testIncorrectTimestamp`. Move this test under `testTimestamp` and add `testIncorrectDatetime` test.

File: plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java (hunk 1 above)
Comment (abusk) [reply plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java]: Yea, typeHandle.getJdbcType() (93) is same for `datetime` and `timestamp`, we do not need to have separate handle for DATETIME type, added a code comment only.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java (hunk 2 above)
Comment (abusk) [reply plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java]: Yea, adding a common roudrips method for datetime and timestamp test.

**Comment by abusk** [comment]
@ebyhr please review the latest PR

File: plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java [hunk 6]
```diff
@@ -379,6 +379,7 @@ public Optional<ColumnMapping> toColumnMapping(ConnectorSession session, Connect
                 TimeType timeType = createTimeType(getTimePrecision(typeHandle.getRequiredColumnSize()));
                 return Optional.of(timeColumnMapping(timeType));
             case Types.TIMESTAMP:
+                //This jdbcType maps both MariaDB data type - TIMESTAMP and DATETIME to Trino TIMESTAMP type.
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java:382]:
```suggestion
                // This jdbcType maps both MariaDB TIMESTAMP and DATETIME types to Trino TIMESTAMP type
```

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java [hunk 7]
```diff
@@ -910,4 +963,9 @@ private static boolean isGap(ZoneId zone, LocalDate date)
     {
         return zone.getRules().getValidOffsets(date.atStartOfDay()).isEmpty();
     }
+
+    private String getInputTypeWithPrecision(String jdbcTypeName, int precision)
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:967]: Inline this method. No need to extract this single line operation.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java [hunk 8]
```diff
@@ -876,6 +835,100 @@ public void testTimestampCoercion()
...
+                .addRoundTrip("datetime(4)", "NULL", createTimestampType(4), "CAST(NULL AS TIMESTAMP(4))")
+                .addRoundTrip("datetime(5)", "NULL", createTimestampType(5), "CAST(NULL AS TIMESTAMP(5))")
+                .addRoundTrip("datetime(6)", "NULL", createTimestampType(6), "CAST(NULL AS TIMESTAMP(6))")
+                .execute(getQueryRunner(), session, mariaDbCreateAndInsert("tpch.test_datetime"));
+    }
+
+    @Test
+    public void testIncorrectDateTime()
+    {
+        //MariaDB support any valid TIMESTAMP literal of type YYYY-MM-DD hh:mm:ss.S(up to 6 digits), fails otherwise.
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:878]:
```suggestion
        // MariaDB supports any valid TIMESTAMP literal of type YYYY-MM-DD hh:mm:ss.S(up to 6 digits), fails otherwise.
```

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java [hunk 9]
```diff
@@ -876,6 +835,100 @@ public void testTimestampCoercion()
...
+                .addRoundTrip("datetime(6)", "TIMESTAMP '5555-12-31 23:59:59.999995'", createTimestampType(6), "TIMESTAMP '5555-12-31 23:59:59.999995'")
+
+                // null
+                .addRoundTrip("datetime(0)", "NULL", createTimestampType(0), "CAST(NULL AS TIMESTAMP(0))")
+                .addRoundTrip("datetime(1)", "NULL", createTimestampType(1), "CAST(NULL AS TIMESTAMP(1))")
+                .addRoundTrip("datetime(2)", "NULL", createTimestampType(2), "CAST(NULL AS TIMESTAMP(2))")
+                .addRoundTrip("datetime(3)", "NULL", createTimestampType(3), "CAST(NULL AS TIMESTAMP(3))")
+                .addRoundTrip("datetime(4)", "NULL", createTimestampType(4), "CAST(NULL AS TIMESTAMP(4))")
+                .addRoundTrip("datetime(5)", "NULL", createTimestampType(5), "CAST(NULL AS TIMESTAMP(5))")
+                .addRoundTrip("datetime(6)", "NULL", createTimestampType(6), "CAST(NULL AS TIMESTAMP(6))")
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:871]: We should test these null cases with `timestamp` type as well.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java [hunk 10]
```diff
@@ -876,6 +835,100 @@ public void testTimestampCoercion()
...
+    public void testDatetime()
+    {
+        testDatetime(UTC);
+        testDatetime(jvmZone);
+        testDatetime(vilnius);
+        testDatetime(kathmandu);
+        testDatetime(TestingSession.DEFAULT_TIME_ZONE_KEY.getZoneId());
+    }
+
+    private void testDatetime(ZoneId sessionZone)
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:848]: Please add min and max roundtrips.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java [hunk 11]
```diff
@@ -876,6 +835,100 @@ public void testTimestampCoercion()
...
+        try (TestTable table = new TestTable(server::execute, "test_incorrect_datetime", "(dt DATETIME)")) {
+            assertQueryFails(format("INSERT INTO %s VALUES (TIMESTAMP '999-01-01 00:00:00.000')", table.getName()), ".*not a valid TIMESTAMP literal.*");
+            assertQueryFails(format("INSERT INTO %s VALUES (TIMESTAMP '10000-01-19 03:14:08.000')", table.getName()), ".*Failed to insert data.*");
+        }
+    }
+
+    private SqlDataTypeTest createCommonDatetimeTimestampRoundTrip(String inputType)
+    {
+        String withPrecision3 = getInputTypeWithPrecision(inputType, 3);
+        String withPrecision6 = getInputTypeWithPrecision(inputType, 6);
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:888]: `inputType + "(3)"` & `inputType + "(6)"` look sufficient. I would inline these variables for consistency with other roundtrips.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java [hunk 12]
```diff
@@ -876,6 +835,100 @@ public void testTimestampCoercion()
...
+        createCommonDatetimeTimestampRoundTrip("datetime")
+                // before epoch
+                .addRoundTrip("datetime(3)", "TIMESTAMP '1958-01-01 13:18:03.123'", createTimestampType(3), "TIMESTAMP '1958-01-01 13:18:03.123'")
+                .addRoundTrip("datetime(6)", "TIMESTAMP '1969-12-31 23:59:59.999995'", createTimestampType(6), "TIMESTAMP '1969-12-31 23:59:59.999995'")
+                .addRoundTrip("datetime(6)", "TIMESTAMP '1969-12-31 23:59:59.999949'", createTimestampType(6), "TIMESTAMP '1969-12-31 23:59:59.999949'")
+                .addRoundTrip("datetime(6)", "TIMESTAMP '1969-12-31 23:59:59.999994'", createTimestampType(6), "TIMESTAMP '1969-12-31 23:59:59.999994'")
+
+                //After epoch
+                .addRoundTrip("datetime(3)", "TIMESTAMP '3555-01-01 13:18:03.123'", createTimestampType(3), "TIMESTAMP '3555-01-01 13:18:03.123'")
+                .addRoundTrip("datetime(6)", "TIMESTAMP '5555-12-31 23:59:59.999995'", createTimestampType(6), "TIMESTAMP '5555-12-31 23:59:59.999995'")
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:862]: Remove as the roundtrips for after epoch already exist in a helper method.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java [hunk 13]
```diff
@@ -876,6 +835,100 @@ public void testTimestampCoercion()
...
+                .addRoundTrip(withPrecision3, "TIMESTAMP '2018-10-28 01:33:17.456'", createTimestampType(3), "TIMESTAMP '2018-10-28 01:33:17.456'")
+                // time double in Vilnius
+                .addRoundTrip(withPrecision3, "TIMESTAMP '2018-10-28 03:33:33.333'", createTimestampType(3), "TIMESTAMP '2018-10-28 03:33:33.333'")
+                .addRoundTrip(withPrecision3, "TIMESTAMP '1970-01-01 00:13:42.000'", createTimestampType(3), "TIMESTAMP '1970-01-01 00:13:42.000'")
+                .addRoundTrip(withPrecision3, "TIMESTAMP '2018-04-01 02:13:55.123'", createTimestampType(3), "TIMESTAMP '2018-04-01 02:13:55.123'")
+                // time gap in Vilnius
+                .addRoundTrip(withPrecision3, "TIMESTAMP '2018-03-25 03:17:17.000'", createTimestampType(3), "TIMESTAMP '2018-03-25 03:17:17.000'")
+                // time gap in Kathmandu
+                .addRoundTrip(withPrecision3, "TIMESTAMP '1986-01-01 00:13:07.000'", createTimestampType(3), "TIMESTAMP '1986-01-01 00:13:07.000'")
+                // max value 2038-01-19 03:14:07
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:902]: This isn't the max value in case of datetime type. Let's mention timestamp type in a comment. 
```
// max value 2038-01-19 03:14:07 in MariaDB timestamp type
```

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java [hunk 14]
```diff
@@ -876,6 +835,100 @@ public void testTimestampCoercion()
...
+    public void testIncorrectDateTime()
+    {
+        //MariaDB support any valid TIMESTAMP literal of type YYYY-MM-DD hh:mm:ss.S(up to 6 digits), fails otherwise.
+        try (TestTable table = new TestTable(server::execute, "test_incorrect_datetime", "(dt DATETIME)")) {
+            assertQueryFails(format("INSERT INTO %s VALUES (TIMESTAMP '999-01-01 00:00:00.000')", table.getName()), ".*not a valid TIMESTAMP literal.*");
+            assertQueryFails(format("INSERT INTO %s VALUES (TIMESTAMP '10000-01-19 03:14:08.000')", table.getName()), ".*Failed to insert data.*");
+        }
+    }
+
+    private SqlDataTypeTest createCommonDatetimeTimestampRoundTrip(String inputType)
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:885]:
```suggestion
    private SqlDataTypeTest timestampRoundTrips(String inputType)
```

**Review by ebyhr — COMMENTED** [review]
Please squash commits into one and fix the commit title as "Test datetime type in MariaDB connector"

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java (hunk 13 above)
Comment (abusk) [reply plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java]: updated

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java (hunk 9 above)
Comment (abusk) [reply plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java]: Added NULL cases for `timestamp` as well.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java (hunk 10 above)
Comment (abusk) [reply plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:854]: Added min and max rountrips for `datetime`

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java (hunk 12 above)
Comment (abusk) [reply plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java]: Helper method has roundtrips till max of epoch, this roundtrips exceeds max of epoch which is supported in `datetime` only. Updated the comments.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java [hunk 15]
```diff
@@ -762,49 +763,13 @@ private void testTimestamp(ZoneId sessionZone)
...
-                .addRoundTrip("timestamp(1)", "TIMESTAMP '2020-09-27 12:34:56.1'", createTimestampType(1), "TIMESTAMP '2020-09-27 12:34:56.1'")
-                .addRoundTrip("timestamp(1)", "TIMESTAMP '2020-09-27 12:34:56.9'", createTimestampType(1), "TIMESTAMP '2020-09-27 12:34:56.9'")
-                .addRoundTrip("timestamp(3)", "TIMESTAMP '2020-09-27 12:34:56.123'", createTimestampType(3), "TIMESTAMP '2020-09-27 12:34:56.123'")
-                .addRoundTrip("timestamp(3)", "TIMESTAMP '2020-09-27 12:34:56.999'", createTimestampType(3), "TIMESTAMP '2020-09-27 12:34:56.999'")
-                .addRoundTrip("timestamp(6)", "TIMESTAMP '2020-09-27 12:34:56.123456'", createTimestampType(6), "TIMESTAMP '2020-09-27 12:34:56.123456'")
-
+                // min value 1970-01-01 00:00:00
+                .addRoundTrip("timestamp(3)", "TIMESTAMP '1970-01-01 00:00:01.000'", createTimestampType(3), "TIMESTAMP '1970-01-01 00:00:01.000'")
+                .addRoundTrip("timestamp(6)", "TIMESTAMP '1970-01-01 00:00:01.000000'", createTimestampType(6), "TIMESTAMP '1970-01-01 00:00:01.000000'")
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:772]: The min value in a comment is different from the roundtrip.

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java [hunk 16]
```diff
@@ -876,6 +841,99 @@ public void testTimestampCoercion()
...
+    {
+        Session session = Session.builder(getSession())
+                .setTimeZoneKey(getTimeZoneKey(sessionZone.getId()))
+                .build();
+        timestampRoundTrips("datetime")
+                // max value 9999-12-31 23:59:59
+                .addRoundTrip("datetime(3)", "TIMESTAMP '9999-12-31 23:59:59.999'", createTimestampType(3), "TIMESTAMP '9999-12-31 23:59:59.999'")
+                .addRoundTrip("datetime(6)", "TIMESTAMP '9999-12-31 23:59:59.999999'", createTimestampType(6), "TIMESTAMP '9999-12-31 23:59:59.999999'")
+                // min value 1000-01-01 00:00:00
+                .addRoundTrip("datetime(3)", "TIMESTAMP '1000-01-01 00:00:00.000'", createTimestampType(3), "TIMESTAMP '1000-01-01 00:00:00.000'")
```
Comment (ebyhr) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:864]: This value isn't the min value in MariaDB as far as I confirmed:
```
create table tpch.test(a datetime);
insert into tpch.test values (TIMESTAMP '0999-12-31 00:00:00.000');
select * from tpch.test; -- 0999-12-31 00:00:00.000
```

Comment (abusk) [reply plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java:864]: Ah it is supporting min value of `0001-01-01 00:00:00`,  i followed this doc https://mariadb.com/kb/en/datetime/#supported-values.

Comment (abusk) [reply plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java]: updated

**Review by ebyhr — APPROVED** [review]
